### ISSUE TITLE
asm2wasm unconditional float conversion trap

### ISSUE THREAD

**Issue by jgravelle-google** [WebAssembly/binaryen#983]
Compiling https://github.com/gcc-mirror/gcc/blob/master/gcc/testsuite/gcc.c-torture/execute/20040831-1.c with -O0 creates the following asm.js:

```js
function _main() {
 var $0 = 0.0, $1 = 0.0, $2 = 0, $cmp = 0, $cmp2 = 0, $cond = 0, $conv = 0, $conv1 = 0, $d = 0.0, $l = 0, $retval = 0, label = 0, sp = 0;
 sp = STACKTOP;
 STACKTOP = STACKTOP + 16|0; if ((STACKTOP|0) >= (STACK_MAX|0)) abortStackOverflow(16|0);
 $retval = 0;
 $d = -12.0; //@line 9 "/s/gcc-torture/execute/20040831-1.c"
 $0 = $d; //@line 10 "/s/gcc-torture/execute/20040831-1.c"
 $cmp = $0 > 1.0E+4; //@line 10 "/s/gcc-torture/execute/20040831-1.c"
 $1 = $d;
 $conv = (~~(($1))>>>0); //@line 10 "/s/gcc-torture/execute/20040831-1.c"
 $conv1 = (~~(($1))); //@line 10 "/s/gcc-torture/execute/20040831-1.c"
 $cond = $cmp ? $conv : $conv1; //@line 10 "/s/gcc-torture/execute/20040831-1.c"
 $l = $cond; //@line 10 "/s/gcc-torture/execute/20040831-1.c"
 $2 = $l; //@line 11 "/s/gcc-torture/execute/20040831-1.c"
 $cmp2 = ($2|0)!=(-12); //@line 11 "/s/gcc-torture/execute/20040831-1.c"
 if ($cmp2) {
  _abort(); //@line 12 "/s/gcc-torture/execute/20040831-1.c"
  // unreachable; //@line 12 "/s/gcc-torture/execute/20040831-1.c"
 } else {
  _exit(0); //@line 13 "/s/gcc-torture/execute/20040831-1.c"
  // unreachable; //@line 13 "/s/gcc-torture/execute/20040831-1.c"
 }
 return (0)|0;
}
```

Running asm2wasm on that turns `$conv` and `$conv1` into `i32.trunc_u` and `i32.trunc_s`, respectively. They both execute unconditionally, and `i32.trunc_u` traps because it can't represent `-12.0`.

The bug is ultimately that the conversions happen unconditionally.

**Comment by kripken** [comment]
Isn't this clang taking advantage of undefined behavior?

I don't think we can do anything in binaryen here. Users can set the trap mode to js or clamp, though, which prevents the trap. (i.e. this now traps by default only because we changed the default trap mode to allow traps).

**Comment by kripken** [comment]
Specifically, conditional conversions become unconditional when LLVM simplifies the CFG (we run that pass even in -O0 as part of legalization).

**Comment by jgravelle-google** [comment]
I think this is really a JSBackend issue, but it's only user-visible with asm2wasm which is why I filed it in Binaryen.
And it isn't clang, the generated IR that we get out of clang still leaves the conversion as being done conditionally:
```llvm
; Function Attrs: noinline nounwind
define i32 @main() #0 !dbg !9 {
entry:
  %retval = alloca i32, align 4
  %d = alloca double, align 8
  %l = alloca i32, align 4
  store i32 0, i32* %retval, align 4
  call void @llvm.dbg.declare(metadata double* %d, metadata !13, metadata !15), !dbg !16
  store double -1.200000e+01, double* %d, align 8, !dbg !16
  call void @llvm.dbg.declare(metadata i32* %l, metadata !17, metadata !15), !dbg !18
  %0 = load double, double* %d, align 8, !dbg !19
  %cmp = fcmp ogt double %0, 1.000000e+04, !dbg !20
  br i1 %cmp, label %cond.true, label %cond.false, !dbg !21

cond.true:                                        ; preds = %entry
  %1 = load double, double* %d, align 8, !dbg !22
  %conv = fptoui double %1 to i32, !dbg !24
  br label %cond.end, !dbg !25

cond.false:                                       ; preds = %entry
  %2 = load double, double* %d, align 8, !dbg !26
  %conv1 = fptosi double %2 to i32, !dbg !28
  br label %cond.end, !dbg !29

cond.end:                                         ; preds = %cond.false, %cond.true
  %cond = phi i32 [ %conv, %cond.true ], [ %conv1, %cond.false ], !dbg !30
  store i32 %cond, i32* %l, align 4, !dbg !32
  %3 = load i32, i32* %l, align 4, !dbg !33
  %cmp2 = icmp ne i32 %3, -12, !dbg !35
  br i1 %cmp2, label %if.then, label %if.end, !dbg !36

if.then:                                          ; preds = %cond.end
  call void @abort() #9, !dbg !37
  unreachable, !dbg !37

if.end:                                           ; preds = %cond.end
  call void @exit(i32 0) #9, !dbg !38
  unreachable, !dbg !38

return:                                           ; No predecessors!
  %4 = load i32, i32* %retval, align 4, !dbg !39
  ret i32 %4, !dbg !39
}
```
The wasm backend leaves it to be conditionally executed (and doesn't trap), and the IR looks pretty much the same for both.

**Comment by kripken** [comment]
Yeah, sorry, see my second comment: it's not clang, it's LLVM's cfg-simplification pass which unconditionalizes it. We do run that in the JS backend, which is why we hit this issue and other backends might not. But isn't simplifying the cfg a valid operation in LLVM at any time? That pass obeys LLVM rules, and LLVM's rules allow unconditionalizing that code because it depends on UB, I believe.

**Comment by dschuff** [comment]
I don't think the code actually has UB. The C ternary operator is short-circuiting, so the conversion to unsigned only gets executed if the operand is > 10000. The 'as-if' rule allows the optimizer to modify the CFG and execute it unconditionally only if the operation has no side effects. I'm guessing that this operation has no side effects on the JS target and other architectures (and is therefore safe), but it does have side effects on wasm, and it's therefore not a valid transformation for the compiler to make.

**Comment by kripken** [comment]
In that case, were we wrong when we said way back that it's ok to make these wasm operations trap, since it's just UB anyhow? To put it another way, you've reached the conclusion that emitting the plain wasm instructions - which might trap - can be wrong in a way that does *not* involve UB. So compilers can't just emit them even if they are ok with ignoring UB.

Concretely, here: what can the JS backend do? It seems like it needs to mark that code as having side effects, so that cfg simplification doesn't unconditionalize it. How would we do that? I hope we don't need to add some new intrinsic function for every wasm math operation.

**Comment by dschuff** [comment]
IIUC the UB in question is converting a float to an unsigned int when the value is negative. Emitting the plain wasm instructions to implement this conversion is always fine, because are allowed to trap when the source program does that. But the source program does not do that. It executes the conversion conditionally. What's not fine is removing the condition.

So I haven't looked at what CFGSimplify or whatever pass does exactly yet, but it must have some way of knowing which actions are side-effecting and which are not (e.g. it wouldn't remove conditions from calls). So if there's some target hook to specify that, then we can hook into it, or if we are comfortable with a more general hack, we may have to do that. But since we haven't seen this with the wasm backend (have we?) I'm thinking there's probably some target-specific behavior we can implement.

**Comment by binji** [comment]
Just curious, why doesn't this fail the same way with the WebAssembly backend? Does it not run the CFGSimplify pass? Or is it already wired into some backend-specific hook?

**Comment by dschuff** [comment]
Yeah I just looked at CFGSimplify and it doesn't look like it does this kind of thing. Are you sure that's the pass that's doing it?

**Comment by dschuff** [comment]
OK, I checked and it is SimplifyCFG that does this, and we just don't happen to run it in the wasm backend. But if you run SimplifyCFG on the wasm triple, you get the same behavior. So this is the same problem  either way.

**Comment by kripken** [comment]
Yeah, looks like we only notice this in the JSBackend because it runs pass for legalization purposes. But in -O1 it would run in both the wasm backend and fastcomp. So this is a general problem. "Luckily", in -O1 other opts get in the way on this testcase the problem does not show up.

But turns out if we make both of those C variables `volatile` then those opts don't interfere and in -O1 we do hit the crash in both fastcomp and the wasm backend.

**Comment by kripken** [comment]
Also seeing this problem on something like `input == 0 ? 0 : int(std::fabs(std::log10(input)))`, where it unconditionally does the log10, fabs, and float-to-int, which can then trap. Might be the same general issue as here, but noting so we remember to check it later.

**Comment by sunfishcode** [comment]
I have a work-in-progress patch [here](https://reviews.llvm.org/D38524) which fixes this in the LLVM wasm backend.
